packages/editor/src/components/autocompleters/user.js

```diff
@@ -48,6 +48,16 @@ export default {
 			</span>,
 		];
 	},
+	allowContext( before, after, textWithoutTrigger ) {
```

Comment:
I took the freedom to risk and experiment a bit by turning the `allowContext` callback into a more general-purpoise filter for the completer, by also passing the `textWithoutTrigger` to it. The previous [version](https://github.com/WordPress/gutenberg/pull/30589/commits/1949a4a23e55930b11101f43bd6933a8908ff39f) was clunkier as it included the logic below in the `Autocomplete` component. 

I think this is a relatively safe change, even if I'm adding to the method signature, but let me know if you think a new function would be better, or if I should fallback to https://github.com/WordPress/gutenberg/pull/30589/commits/1949a4a23e55930b11101f43bd6933a8908ff39f.

Reply:
See https://github.com/WordPress/gutenberg/pull/30589#issuecomment-815517413 where I left some more thought. I think we need something general like the maximum number of chars that gets processed and a way to fine-tune for individual completers like you propose 👍🏻

Reply:
TBH I don't love the idea of bending `allowContext` to suit our needs here 😛 
OTOH, I also don't love the idea of a customizable max number of chars for individual completers -- I don't think that reflects the behavior of completion as I'd imagine it.

I'll leave a top-level comment with some more discussion.